# Next-Gen Sequence Analysis Workshop (2017)

This is the schedule for the
[ANGUS workshop at UC Davis](http://ivory.idyll.org/dibsi/ANGUS.html).
which will run from ...

See [the master schedule](http://ivory.idyll.org/dibsi/SCHEDULE.html) for
times etc; this information will be integrated below soon.

## The main workshop materials

### Monday, Day 1: Introduction

Introductions!

[Booting a cloud computer from Jetstream!](jetstream/boot.html)

[Running BLAST at the command line](running-command-line-blast.html)

Assessment.

Student 30-second introductions.

### Tuesday, Day 2: Command line.

[Running large scale BLASTs and output.](running-blast-large-scale.html)

[Visualizing BLAST score distributions in RStudio](visualizing-blast-scores-with-RStudio.html)

[Review and explore: Command line UNIX, and R/RStudio](command-line-and-rstudio.html)

Evening: student presentations and questions!

### Wednesday, Day 3: Mapping, aligning, and sequence files.

BWA and samtools and variant calling.

Introduction to automation.

(Evening: free time)

### Thursday, Day 4: Reference independent analyses: assembly and k-mers

Microbial genome assembly.

Long reads and short reads.

Evening: k-mers and k-mer-based analyses; sourmash.

### Friday, Day 5: Counting and differential expression (round 1).

Variants, RNAseq, differential foo.

### Saturday, Day 6: Automation and repeatability.

(Morning: free time)

Afternoon: more shell, loops, and scripting. (SC shell lesson.)

### Sunday, Day 7: Day of rest.

### Monday, Day 8: TBD.

### Tuesday, Day 9: TBD.

### Wednesday, Day 10: TBD.

### Thursday, Day 11: RNAseq

### Friday, Day 12: RNAseq.

Asessment.

### Saturday, last day: Summary

Closing talk.

Further resources.

Questions and answers.

Course review, input, and feedback.

## Other material

[Instructor technical info](for-instructors/index)
