# Next-Gen Sequence Analysis Workshop (2017)

These are the schedule and classroom materials for the
[ANGUS workshop at UC Davis](http://ivory.idyll.org/dibsi/ANGUS.html),
which will run from June 26th to July 8th, 2017.

Twitter hash tag: [#ngs2017](https://twitter.com/search?f=tweets&q=%23ngs2017&src=typd)

## The main workshop materials

### Monday, Day 1: Introduction

7:30pm - 9pm, Valley Hall.

Introductions!

[Booting a cloud computer from Jetstream!](jetstream/boot.html)

[Running BLAST at the command line](running-command-line-blast.html)

Assessment.

Student 30-second introductions.

### Tuesday, Day 2: Command line.

9am lecture: C. Titus Brown, UC Davis - "On Biology and Data Analysis."

Morning: 10:15am.
* [Running large scale BLASTs and output.](running-blast-large-scale.html)

Afternoon: 1:15am - 3pm.
* [Visualizing BLAST score distributions in RStudio](visualizing-blast-scores-with-RStudio.html)
* [Review and explore: Command line UNIX, and R/RStudio](command-line-and-rstudio.html)

Evening: 7:30pm-9pm: student presentations and questions! Ice cream social?

### Wednesday, Day 3: Mapping, aligning, and sequence files.

9am lecture:
[Melissa Wilson Sayres](http://www.wilsonsayreslab.org/), Arizona State University - "Sex-biased genome evolution"

Morning and afternoon: 10:15am-4pm.
* [Quality trimming your reads.](quality-trimming.html)
* [Mapping and samtools and variant calling.](variant-calling.html)
* [Introduction to automation.](introduction-to-automation.html)

(Evening: free time - [Wed Farmers' market!](http://www.davisfarmersmarket.org/))

### Thursday, Day 4: Introduction to counting and differential expression.

9am lecture: [Chris Hamm](https://butterflyology.github.io/about-me.html), Monsanto - "How to design a good experiment"

Morning: 10:15am-noon
* Introduction to R (guest instructor for one room: Tracy Teal)

Afternoon and evening: 1:15pm-3pm, 7:30-9pm.
* [Short reads and counting for variant calling and differential expression](counting.html)
* Variants, RNAseq, & differential expression analysis.

### Friday, Day 5: Reference independent analyses: assembly and k-mers

9am lecture: [Adrienne Roeder](http://roeder.wicmb.cornell.edu/), Cornell - Reaching biological conclusions from RNA-seq: the good, the bad, and the ugly

Morning: 10:15am-noon.
* [Microbial genome assembly.](genome-assembly.html)
* Annotating assemblies w/prokka; BLAST and pangenome analysis.

Afternoon: 1:15pm-4pm.
* [Reference independent analyses with k-mers; sourmash.](kmers-and-sourmash.html)

### Saturday, Day 6: Automation and repeatability.

(Morning: free time / Saturday market)

Afternoon (1-4pm): [More shell, loops, and scripting.](automation-2.html)

### Sunday, Day 7: Day of rest!

### Monday, Day 8: Genomes, long reads, and GWAS.

9am lecture: [Erich Schwarz](https://mbg.cornell.edu/people/erich-schwarz), Cornell - "Assembling and biologically interpreting nematode genomes"

Morning and afternoon:
* [Long reads and short reads.](long-and-short-reads.html) (Lisa Cohen)
* Variant calling and big genomes: GTAK. (Tamer Mansour)

### Tuesday, Day 9: TBD.

9am-noon: lecture/tutorials/panel discussion.

afternoon/evening: [July 4th celebration at Community Park](http://cityofdavis.org/city-hall/city-manager-s-office/community-events/fourth-of-july)

### Wednesday, Day 10: TBD.

9am lecture: [Megan Dennis](http://www.dennislab.org/), UC Davis - "Complex genomic variation and its role in human evolution and disease"

Morning and afternoon:
* Genome Wide Association Studies and PLINK. (Tamer Mansour)

### Thursday, Day 11: RNAseq

9am lecture: [Michael I Love](https://mikelove.github.io/), UNC Chapel Hill - "Statistics and bias correction in RNAseq differential expression analysis"

### Friday, Day 12: RNAseq.

9am lecture: [Robert Patro](http://www.robpatro.com/redesign/), Stony Brook University - "Don't count on it: Pragmatic and theoretical concerns and best practices for mapping and quantifying RNA-seq data"

Assessment.

### Saturday, last day: Summary

9am lecture: C. Titus Brown, UC Davis - "Effectively infinite: next steps in Data Intensive Biology."

Further resources.

Questions and answers.

Course review, input, and feedback.

## Other material

[Instructor technical info](for-instructors/index)
